File: StringIterator.cpp

#include<string>
using namespace std;

class StringIterator {
public:

    string input;
    char letter;
    int index;
    int frequencyLetter;
    const char SPACE_CHAR = ' ';

    StringIterator(string compressedString) {
        input = compressedString;
        frequencyLetter = 0;
        index = 0;
    }

    char next() {
        if (!hasNext()) {
            return SPACE_CHAR;
        }

        if (frequencyLetter == 0) {
            letter = input[index++];
            while (index < input.size() && isdigit(input[index])) {
                frequencyLetter = frequencyLetter * 10 + (input[index++] - '0');
            }
        }

        frequencyLetter--;
        return letter;
    }

    bool hasNext() {
        return index < input.size() || frequencyLetter > 0;
    }
};

File: StringIterator.js

/**
 * @param {string} compressedString
 */
var StringIterator = function (compressedString) {
    this.input = compressedString.split('');
    this.index = 0;
    this.frequencyLetter = 0;
    this.letter = ' ';
    this.SPACE_CHAR = ' ';
};

/**
 * @return {character}
 */
StringIterator.prototype.next = function () {

    if (!this.hasNext()) {
        return this.SPACE_CHAR;
    }

    if (this.frequencyLetter === 0) {
        this.letter = this.input[this.index++];
        while (this.index < this.input.length && !isNaN(this.input[this.index])) {
            this.frequencyLetter = this.frequencyLetter * 10 + (this.input[this.index++] - '0');
        }
    }

    this.frequencyLetter--;
    return this.letter;
};

/**
 * @return {boolean}
 */
StringIterator.prototype.hasNext = function () {
    return this.index < this.input.length || this.frequencyLetter > 0;
};
